flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c

```diff
@@ -0,0 +1,603 @@
+}
+
+static void _msp_send_attitude(struct msp_bridge *m)
+{
+	union {
+		uint8_t buf[0];
+		struct {
+			int16_t x;
+			int16_t y;
+			int16_t h;
+		} att;
+	} data;
+	AttitudeActualData attActual;
+
+	AttitudeActualGet(&attActual);
+
+	data.att.x = attActual.Roll * -10;
+	data.att.y = attActual.Pitch * 10;
+	data.att.h = 0 * 10;
+
+	msp_send(m, MSP_ATTITUDE, data.buf, sizeof(data));
+}
+
+static void _msp_send_status(struct msp_bridge *m)
+{
+	union {
+		uint8_t buf[0];
+		struct {
+			uint16_t cycleTime;
+			uint16_t i2cErrors;
+			uint16_t sensors;
+			uint32_t flags;
+			uint8_t setting;
+		} __attribute__((packed)) status;
+	} data;
+	// TODO: https://github.com/TauLabs/TauLabs/blob/next/shared/uavobjectdefinition/actuatordesired.xml#L8
+	data.status.cycleTime = 0;
+	data.status.i2cErrors = 0;
+	data.status.sensors = MSP_SENSOR_ACC; // TODO fill in
+	data.status.flags = 0;
+	data.status.setting = 0;
+
+	if (FlightStatusHandle() != NULL) {
+		FlightStatusData flight_status;
+		FlightStatusGet(&flight_status);
+
+		data.status.flags = flight_status.Armed == FLIGHTSTATUS_ARMED_ARMED;
+
+		for (int i = 1; msp_boxes[i].mode != MSP_BOX_LAST; i++) {
+			if (flight_status.FlightMode == msp_boxes[i].tlmode) {
+				data.status.flags |= (1 << i);
+			}
+		}
+	}
+
+	msp_send(m, MSP_STATUS, data.buf, sizeof(data));
+}
+
+static void _msp_send_analog(struct msp_bridge *m)
+{
+	union {
+		uint8_t buf[0];
+		struct {
+			uint8_t vbat;
+			uint16_t powerMeterSum;
+			uint16_t rssi;
+			uint16_t current;
+		} __attribute__((packed)) status;
+	} data;
+	data.status.powerMeterSum = 0;
+
+	FlightBatterySettingsData batSettings = {};
+	FlightBatteryStateData batState = {};
+
+	if (FlightBatteryStateHandle() != NULL)
+		FlightBatteryStateGet(&batState);
+	if (FlightBatterySettingsHandle() != NULL) {
+		FlightBatterySettingsGet(&batSettings);
+	}
+
+	// TODO:  Verify these with a board that has working power stuff
+	if (batSettings.VoltagePin != FLIGHTBATTERYSETTINGS_VOLTAGEPIN_NONE)
+		data.status.vbat = (uint8_t)lroundf(batState.Voltage * 10);
+
+	if (batSettings.CurrentPin != FLIGHTBATTERYSETTINGS_CURRENTPIN_NONE)
+		data.status.current = lroundf(batState.Current * 10);
+
+	ManualControlCommandData manualState;
+	ManualControlCommandGet(&manualState);
+
+	// MSP RSSI's range is 0-1023
+	data.status.rssi = (manualState.Rssi >= 0 && manualState.Rssi <= 100) ? manualState.Rssi * 10 : 0;
+
+	msp_send(m, MSP_ANALOG, data.buf, sizeof(data));
+}
+
+static void _msp_send_ident(struct msp_bridge *m)
+{
+	// TODO
+}
+
+static void _msp_send_raw_gps(struct msp_bridge *m)
+{
+	// TODO
+}
+
+static void _msp_send_comp_gps(struct msp_bridge *m)
+{
+	// TODO
+}
+
+static void _msp_send_altitude(struct msp_bridge *m)
+{
+	// TODO
+}
+
+static void _msp_send_channels(struct msp_bridge *m)
+{
+	ManualControlCommandData manualState;
+	ManualControlCommandGet(&manualState);
+
+	// MSP RC order is Roll/Pitch/Yaw/Throttle/AUX1/AUX2/AUX3/AUX4
+	int channelMap[] = {
```

Comment:
I am embarrassed for the compiler that this is necessary, but static const uint8_t please?

It's putting it on stack:

```
 843 022a 0FCF                  ldmia   r7!, {r0, r1, r2, r3}
 844 022c 05AE                  add     r6, sp, #20
 845 022e 0FC6                  stmia   r6!, {r0, r1, r2, r3}
 846 0230 97E80F00              ldmia   r7, {r0, r1, r2, r3}
 847 0234 86E80F00              stmia   r6, {r0, r1, r2, r3}
```

and using 4 bytes per index of flash:

```
 1032                   .LC0:
 1033 0000 01000000             .word   1
 1034 0004 02000000             .word   2
 1035 0008 03000000             .word   3
 1036 000c 00000000             .word   0
 1037 0010 04000000             .word   4
 1038 0014 06000000             .word   6
 1039 0018 07000000             .word   7
 1040 001c 08000000             .word   8
```